### ISSUE TITLE
[Help] <title>

### ISSUE THREAD

**Issue by YefZhao** [THUDM/ChatGLM-6B#1058]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

p-tuning v2文章里说把prefix加在transformer第21至24层效果比较好，如图。
![image](https://github.com/THUDM/ChatGLM-6B/assets/58016897/03968463-9945-464a-9abe-ee720bb1dfe5)
请问本仓库p-tuning实现的是把prefix加在21-24层吗？这个设置是在代码的那一部分呢？

### Expected Behavior

答疑

### Steps To Reproduce

无

### Environment

```markdown
OS: Ubuntu 20.04
Python: 3.8
Transformers: 4.26.1
PyTorch: 1.12
CUDA Support: True
```

**Comment by runzhi214** [comment]
根据transformer pretrained mode的api：https://huggingface.co/docs/transformers/main_classes/model
在p-tuning的main.py中有这个方法：
```        
model.gradient_checkpointing_enable()
model.enable_input_require_grads()
```
enable_input_require_grads这个方法是启动模型input embedding层的梯度。
所以我理解是p-tuning调整的是input_embedding?

**Comment by runzhi214** [comment]
[https://github.com/huggingface/transformers/blob/a7920065f2cfd2549b838f9a30afd7c265fcdd88/src/transformers/modeling_utils.py#L1185](url)
>     def enable_input_require_grads(self):
        """
        Enables the gradients for the input embeddings. This is useful for fine-tuning adapter weights while keeping
        the model weights fixed.
        """

        def make_inputs_require_grads(module, input, output):
            output.requires_grad_(True)

        self._require_grads_hook = self.get_input_embeddings().register_forward_hook(make_inputs_require_grads)

**Comment by YefZhao** [comment]
> 根据transformer pretrained mode的api：https://huggingface.co/docs/transformers/main_classes/model
> 
> Enables the gradients for the input embeddings. This is useful for fine-tuning adapter weights while keeping the model weights fixed.
> 
> 在p-tuning的main.py中有这个方法：
> 
> ```
> model.gradient_checkpointing_enable()
> model.enable_input_require_grads()
> ```
> 
> enable_input_require_grads这个方法是启动模型input embedding层的梯度。 所以我理解是p-tuning调整的是input_embedding?

感谢回复，p-tuning的确是微调一个pre_seq_len长度的continuous prompts，加在transformer的一层或多层。论文说加在更深的层中效果更好，所以不确定这里的实现是加在哪些层，以及在什么地方设置。NLP萌新，理解不当麻烦指教。

**Comment by echoht** [comment]
> 

看源码，实现在所有层了。
源码位置在modeling_chatglm.py 136行，和ptuning v2，代码一致。
![企业微信截图_16844866285937](https://github.com/THUDM/ChatGLM-6B/assets/48375360/819154ad-ee3e-4b71-899e-5a916e847cf8)

**Comment by created-Bi** [comment]
> > 
> 
> 看源码，实现在所有层了。 源码位置在modeling_chatglm.py 136行，和ptuning v2，代码一致。 ![企业微信截图_16844866285937](https://user-images.githubusercontent.com/48375360/239488643-819154ad-ee3e-4b71-899e-5a916e847cf8.png)

您好，多问一句，PrefixEncoder的forward函数的输入是什么？我看论文中的图应该指的是pre_seq_len的tokens作为输入？

**Comment by shimudong** [comment]
> > > 
> > 
> > 
> > 看源码，实现在所有层了。 源码位置在modeling_chatglm.py 136行，和ptuning v2，代码一致。 ![企业微信截图_16844866285937](https://user-images.githubusercontent.com/48375360/239488643-819154ad-ee3e-4b71-899e-5a916e847cf8.png)
> 
> 您好，多问一句，PrefixEncoder的forward函数的输入是什么？我看论文中的图应该指的是pre_seq_len的tokens作为输入？

PrefixEncoder的输入是固定的

`torch.arange(self.pre_seq_len).long()`
